crypto/rsa/rsa_gen.c

```diff
@@ -53,19 +64,33 @@ static int rsa_builtin_keygen(RSA *rsa, int bits, BIGNUM *e_value,
         goto err;
     }
 
+    if (primes < RSA_DEFAULT_PRIME_NUM
+        || primes > RSA_MAX_PRIME_NUM || bits <= primes) {
+        RSAerr(RSA_F_RSA_BUILTIN_KEYGEN, RSA_R_KEY_PRIME_NUM_INVALID);
+        goto err;
+    }
+
     ctx = BN_CTX_new();
     if (ctx == NULL)
         goto err;
     BN_CTX_start(ctx);
     r0 = BN_CTX_get(ctx);
     r1 = BN_CTX_get(ctx);
     r2 = BN_CTX_get(ctx);
-    r3 = BN_CTX_get(ctx);
-    if (r3 == NULL)
+    if (r2 == NULL)
         goto err;
 
-    bitsp = (bits + 1) / 2;
-    bitsq = bits - bitsp;
+    /* divide bits into 'primes' pieces evenly */
+    quo = bits / primes;
+    rmd = bits % primes;
+    for (i = 0; i < primes; i++) {
+        bitsr[i] = (i < rmd) ? quo + 1 : quo;
+        if (bitsr[i] < RSA_MIN_PRIME_SIZE) {
```

Comment:
And RSA_MIN_PRIME_SIZE is defined as 8. 8 *bits*?!  I for one would actually argue that it should be much more restrictive...

Reply:
Errr, how about 64? same as normal RSA proposed here: #4547

Reply:
Well, I for *one* would go even further [here and elsewhere] with rationale that setting some more sensible limits is as educational as setting that low limit. I mean I suppose the purpose of allowing 64 is to keep it feasible to perform operations on paper to compare to results generated by OpenSSL. But is it actually meaningful exercise nowadays? By the time student comes to the point of grasping RSA I bet she is actually expected to have some kind of programming skills. And so forcing student to put together a bignum program say in Python is actually more meaningful exercise. Well, but once again, it's me, or in other words let's here other opinions... But in either case we seem to agree that 8 is unacceptable.

Reply:
Yes. The main job of this limitation is to not get stuck on getting different prime number when generating too small key with a large prime number, say a 16-prime-64-bit key...

So if 64 is used here, along with the maximum prime number of 16, the smallest key of 16 primes will be 1024 bit and that would be acceptable.

Another choice is to remove this checking in the key generation function, and it's the user's  duty to make sure the each prime is too small to get infinity loop in generating primes...

Reply:
> smallest key of 16 primes will be 1024 bit and that would be acceptable.

Acceptable for what? For inadequate security?

> user's duty

Speaking of which. It's probably appropriate to at least mention that number of primes is a security trade-off, i.e. more primes give worse security.

Reply:
> Acceptable for what? For inadequate security?

Yes if someone doesn't know that 1024-bit key is a 16-prime key...

> Speaking of which. It's probably appropriate to at least mention that number of primes is a security trade-off, i.e. more primes give worse security.

In document?

Reply:
Once again, I for *one* find 64 bits lower limit for prime factor unreasonably low. We need more opinions.

> > more primes give worse security.

> In document?

Yes, naturally.

Reply:
> We need more opinions.

That's off-topic for this PR, should be discussed here: #4547

Reply:
Formally speaking these are two different issues. #4547 discusses public key length, while what we are talking about here is private factors' lengths. To illuminate the difference one can formulate the question as following. How many [excess] primes can one use without breaking security promise for given public key length. Well, when put this way one can even question adequacy of mechanical "no more than 16 primes" and "minimum factor length" checks. I mean formally speaking the number of primes should depend on public key length, and should not be subject to arbitrary slicing.

Reply:
#4547 is a more proper place to discuss even if these are two issues...and that is because:

The motivation of this limitation was not about security, but only for not going into infinite loops if the prime length is too small while finding new primes at a given length. And also there are *no* public key length limitations for normal 2-prime RSA neither (the `< 16` above is also used to avoid infinite loops but not the security reasons)

Another choice for security of multi-prime RSA, we could just provide the 'mechanism' (e.g. the functionality) not the 'strategy' (e.g. how to use it safely), and leave the 'strategy' part to the users. If a user just wants a very insecure 1024 bits with 16-primes key, we give them that. So if we do nothing other than avoid infinite loops, that would be also acceptable.

For the security issue of multi-prime RSA, that's more complicated since it depends on two factors - the key length and the number of primes - as you mentioned. And that's right. And this is also why I think $4547 is more proper. The netter action for this PR now is to put some sentences into docs to describe 'for a specific key length, the more the number of primes, the weaker the security strength', otherwise there would be another hundreds of comments in this PR addressing a wider topic (security of multi-prime RSA) other than the functionality of multi-prime RSA itself...